### PR TITLE
pdf dynamic inventory columns

### PR DESCRIPTION
**Pull request by rafalpospiech**

### CODE REVIEW

File: modules/Vtiger/models/PDF.php [hunk 1]
```diff
@@ -538,4 +552,53 @@ public static function zipAndDownload(array $fileNames)
...
+	 * Get column scheme for specified record.
+	 *
+	 * @param int|string $recordId
+	 *
+	 * @return array
+	 */
+	public static function getColumnsForRecord($recordId, string $moduleName)
+	{
+		$columnsJSON = (new App\Db\Query())
+			->select('columns')
```
Comment (rskrzypczak) [inline modules/Vtiger/models/PDF.php]:
```suggestion
			->select(['columns'])
```

File: modules/Vtiger/models/PDF.php [hunk 2]
```diff
@@ -538,4 +552,53 @@ public static function zipAndDownload(array $fileNames)
...
+			->from('a_yf_pdf_inv_scheme')
+			->where(['crmid' => $recordId])
+			->createCommand()
```
Comment (rskrzypczak) [inline modules/Vtiger/models/PDF.php]:
```suggestion
			->scalar();
```

File: modules/Vtiger/models/PDF.php [hunk 3]
```diff
@@ -538,4 +552,53 @@ public static function zipAndDownload(array $fileNames)
...
+			->queryScalar();
```
Comment (rskrzypczak) [inline modules/Vtiger/models/PDF.php]:
```suggestion
```

File: modules/Vtiger/models/PDF.php [hunk 4]
```diff
@@ -538,4 +552,53 @@ public static function zipAndDownload(array $fileNames)
...
+	{
+		$availableColumns = array_keys(Vtiger_Inventory_Model::getInstance($moduleName)->getFields());
+		$invalid = array_diff($columns, $availableColumns);
+		if ($invalid) {
+			throw new \App\Exceptions\IllegalValue('ERR_NOT_ALLOWED_VALUE', 406);
+		}
+		$db = \App\Db::getInstance();
+		$schemeExists = (new \App\Db\Query())
+			->select('crmid')
+			->from('a_yf_pdf_inv_scheme')
```
Comment (rskrzypczak) [inline modules/Vtiger/models/PDF.php]:
```suggestion
			->from('a_#__pdf_inv_scheme')
```

File: modules/Vtiger/models/PDF.php [hunk 5]
```diff
@@ -538,4 +552,53 @@ public static function zipAndDownload(array $fileNames)
...
+
+	/**
+	 * Save column scheme for specified record.
+	 *
+	 * @param int|string $recordId
+	 * @param array      $columns
+	 *
+	 * @throws \App\Exceptions\IllegalValue
+	 */
+	public static function saveColumnsForRecord($recordId, string $moduleName, array $columns)
```
Comment (rskrzypczak) [inline modules/Vtiger/models/PDF.php]: ?

File: modules/Settings/PDF/actions/ColumnScheme.php [hunk 6]
```diff
@@ -0,0 +1,43 @@
+<?php
+
+/**
+ * SaveColumnScheme Action Class for PDF Settings.
+ *
+ * @copyright YetiForce Sp. z o.o
+ * @license YetiForce Public License 3.0 (licenses/LicenseEN.txt or yetiforce.com)
+ * @author Rafal Pospiech <r.pospiech@yetiforce.com>
+ */
+class Settings_PDF_ColumnScheme_Action extends Settings_Vtiger_Index_Action
```
Comment (rskrzypczak) [inline modules/Settings/PDF/actions/ColumnScheme.php]: ?

File: install/install_schema/scheme.sql [hunk 7]
```diff
@@ -160,6 +160,15 @@ CREATE TABLE `a_yf_pdf` (
   KEY `module_name_2` (`module_name`)
 ) ENGINE=InnoDB AUTO_INCREMENT=36 DEFAULT CHARSET=utf8;
 
+/*Table structure for table `a_yf_pdf_inv_col_scheme` */
+
+CREATE TABLE `a_yf_pdf_inv_scheme` (
+  `crmid` INT(10) NOT NULL,
+  `columns` TEXT DEFAULT NULL,
+  KEY `crmid` (`crmid`),
+  CONSTRAINT `crmid` FOREIGN KEY (`crmid`) REFERENCES `vtiger_crmentity` (`crmid`) ON DELETE CASCADE
```
Comment (rskrzypczak) [inline install/install_schema/scheme.sql]: ?

File: modules/Vtiger/actions/PDF.php [hunk 8]
```diff
@@ -236,4 +240,26 @@ public function hasValidTemplate(\App\Request $request)
...
 	}
+
+	/**
+	 * Save inventory column scheme.
+	 *
+	 * @param App\Request $request
+	 */
+	public function saveInventoryColumnScheme(App\Request $request)
+	{
+		$moduleName = $request->get('module', 'String');
```
Comment (rskrzypczak) [inline modules/Vtiger/actions/PDF.php]: String?

File: .editorconfig [hunk 9]
```diff
@@ -8,3 +8,10 @@ end_of_line = lf
 insert_final_newline = true
 trim_trailing_whitespace = true
 max_line_length = 120
+
+[/languages/**]
```
Comment (mariuszkrzaczkowski) [inline .editorconfig]: ?

File: app/TextParser.php [hunk 10]
```diff
@@ -325,7 +340,7 @@ public function setSourceRecord($record, $moduleName = false, $recordModel = fal
 	 */
 	public function setContent($content)
 	{
-		$this->rawContent = $this->content = str_replace('%20%3A%20', ' : ', $content);
+		$this->rawContent = $this->content = str_replace('&nbsp;', ' ', str_replace('%20%3A%20', ' : ', $content));
```
Comment (mariuszkrzaczkowski) [inline app/TextParser.php]: ?

File: app/Pdf/YetiForcePDF.php [hunk 11]
```diff
@@ -113,6 +113,13 @@ class YetiForcePDF extends PDF
 	 */
 	protected $moduleModel;
 
+	/**
+	 * Inventory columns for text parser.
+	 *
+	 * @var array
+	 */
+	protected $inventoryColumns = [];
```
Comment (mariuszkrzaczkowski) [inline app/Pdf/YetiForcePDF.php]:
```suggestion
	protected $params = [];
```

File: app/Pdf/YetiForcePDF.php [hunk 12]
```diff
@@ -328,6 +335,7 @@ public function parseVariables(string $str)
 		$textParser = \App\TextParser::getInstanceById($this->recordId, $this->moduleName);
 		$textParser->setType('pdf');
 		$textParser->setParams(['pdf' => $this->moduleModel]);
+		$textParser->setInventoryColumns($this->getInventoryColumns());
```
Comment (mariuszkrzaczkowski) [inline app/Pdf/YetiForcePDF.php]: $textParser->setParams(['pdf' => $this->moduleModel], 'inventoryColumns' =>$this->getInventoryColumns());

File: app/Pdf/YetiForcePDF.php [hunk 13]
```diff
@@ -677,4 +686,27 @@ public function export($recordId, $moduleName, $templateId, $filePath = '', $sav
 	{
 		$this->generateContent($recordId, $moduleName, $templateId, $recordId)->output($filePath, $saveFlag);
 	}
+
```
Comment (mariuszkrzaczkowski) [inline app/Pdf/YetiForcePDF.php]: ?

File: app/TextParser.php [hunk 14]
```diff
@@ -185,6 +186,20 @@ class TextParser
 	 */
 	public $isHtml = true;
 
+	/**
+	 * Inventory columns.
+	 *
+	 * @var array
+	 */
+	protected $inventoryColumns = [];
```
Comment (mariuszkrzaczkowski) [inline app/TextParser.php]: ?

File: app/TextParser.php [hunk 15]
```diff
@@ -1526,4 +1542,27 @@ protected function getInventoryTable(array $config): string
 		}
 		return $html;
 	}
+
+	/**
+	 * Get inventory columns.
+	 *
+	 * @return array
+	 */
+	public function getInventoryColumns()
```
Comment (mariuszkrzaczkowski) [inline app/TextParser.php]: ?

File: app/TextParser.php [hunk 16]
```diff
@@ -1526,4 +1542,27 @@ protected function getInventoryTable(array $config): string
...
+	}
+
+	/**
+	 * Set inventory columns.
+	 *
+	 * @param array $inventoryColumns Inventory columns.
+	 *
+	 * @return self
+	 */
+	public function setInventoryColumns(array $inventoryColumns)
```
Comment (mariuszkrzaczkowski) [inline app/TextParser.php]: ?

File: app/TextParser/DynamicInventoryColumnsTable.php [hunk 17]
```diff
@@ -0,0 +1,43 @@
+<?php
+
```
Comment (mariuszkrzaczkowski) [inline app/TextParser/DynamicInventoryColumnsTable.php]: doc

File: modules/Settings/PDF/models/Module.php [hunk 18]
```diff
@@ -154,4 +154,24 @@ public function getTemplatesByModule($moduleName)
...
+	 * Get template type.
+	 *
+	 * @param Vtiger_PDF_Model $template
+	 *
+	 * @return int template type
+	 */
+	public static function getTemplateType(Vtiger_PDF_Model $template)
+	{
+		$matches = [];
+		$content = str_replace('&nbsp;', ' ', $template->get('body_content'));
```
Comment (mariuszkrzaczkowski) [inline modules/Settings/PDF/models/Module.php]: ?

File: modules/Vtiger/actions/PDF.php (hunk 8 above)
Comment (mariuszkrzaczkowski) [reply modules/Vtiger/actions/PDF.php]:
```suggestion
		$moduleName = $request->getModule();
```

File: modules/Vtiger/actions/PDF.php [hunk 19]
```diff
@@ -236,4 +240,26 @@ public function hasValidTemplate(\App\Request $request)
...
+		$records = $request->getArray('records', 'Integer');
+		$columns = $request->getArray('inventoryColumns', 'String');
+		foreach ($records as $crmId) {
+			Vtiger_PDF_Model::saveInventoryColumnsForRecord($crmId, $moduleName, $columns);
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/actions/PDF.php]: Batch Insert

File: modules/Vtiger/actions/PDF.php [hunk 20]
```diff
@@ -236,4 +240,26 @@ public function hasValidTemplate(\App\Request $request)
...
+		}
+		$output = [
+			'message' => \App\Language::translate('LBL_SCHEME_SAVED', 'Settings:PDF'),
+			'records' => $records
+		];
+		$response = new Vtiger_Response();
+		$response->setResult($output);
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/actions/PDF.php]:
```suggestion
		$response->setResult([
			'message' => \App\Language::translate('LBL_SCHEME_SAVED', 'Settings:PDF'),
			'records' => $records
		]);
```

File: modules/Vtiger/models/PDF.php [hunk 21]
```diff
@@ -471,15 +485,18 @@ public function getBody()
...
-	 * @param string $saveFlag   - save option flag
+	 * @param int    $recordId         - id of a record
+	 * @param string $moduleName       - name of records module
+	 * @param int    $templateId       - id of pdf template
+	 * @param string $filePath         - path name for saving pdf file
+	 * @param string $saveFlag         - save option flag
+	 * @param array  $inventoryColumns
 	 */
-	public static function exportToPdf($recordId, $moduleName, $templateId, $filePath = '', $saveFlag = '')
+	public static function exportToPdf($recordId, $moduleName, $templateId, $filePath = '', $saveFlag = '', array $inventoryColumns = [])
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/models/PDF.php]:
```suggestion
	public static function exportToPdf($recordId, $moduleName, $templateId, $filePath = '', $saveFlag = '', array $params = [])
```

File: modules/Vtiger/models/PDF.php [hunk 22]
```diff
@@ -471,15 +485,18 @@ public function getBody()
...
 	{
-		(new \App\Pdf\YetiForcePDF())->export($recordId, $moduleName, $templateId, $filePath, $saveFlag);
+		(new \App\Pdf\YetiForcePDF())
+			->setInventoryColumns($inventoryColumns)
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/models/PDF.php]:
```suggestion
			->setParams($params)
```

File: modules/Vtiger/models/PDF.php [hunk 23]
```diff
@@ -538,4 +555,52 @@ public static function zipAndDownload(array $fileNames)
 			unlink($file);
 		}
 	}
+
+	/**
+	 * Get column scheme for specified record.
+	 *
+	 * @param int|string $recordId
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/models/PDF.php]: doc

File: modules/Vtiger/models/PDF.php [hunk 24]
```diff
@@ -538,4 +555,52 @@ public static function zipAndDownload(array $fileNames)
...
+		if ($columnsJSON) {
+			return \App\Json::decode($columnsJSON);
+		}
+		return array_keys(Vtiger_Inventory_Model::getInstance($moduleName)->getFields());
+	}
+
+	/**
+	 * Save column scheme for specified record.
+	 *
+	 * @param int|string $recordId
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/models/PDF.php]: doc

File: modules/Vtiger/models/PDF.php [hunk 25]
```diff
@@ -538,4 +555,52 @@ public static function zipAndDownload(array $fileNames)
...
+	 * @param array      $columns
+	 *
+	 * @throws \App\Exceptions\IllegalValue
+	 */
+	public static function saveInventoryColumnsForRecord($recordId, string $moduleName, array $columns)
+	{
+		$availableColumns = array_keys(Vtiger_Inventory_Model::getInstance($moduleName)->getFields());
+		$invalid = array_diff($columns, $availableColumns);
+		if ($invalid) {
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/models/PDF.php]:
```suggestion
		if (array_diff($columns, $availableColumns)) {
```

File: layouts/basic/modules/Vtiger/ExportPDF.tpl [hunk 26]
```diff
@@ -26,49 +26,67 @@
...
+										<select class="select2" name="inventoryColumns[]" multiple="multiple" data-select-cb="registerSelectSortable" data-js="select2 | sortable">
+										{foreach from=$ALL_INVENTORY_COLUMNS item=$LABEL key=$NAME}
+											<option value="{$NAME}"{if in_array($NAME,$SELECTED_INVENTORY_COLUMNS)} selected="selected"{/if}>{\App\Language::translate($LABEL, $MODULE_NAME)}</option>
+										{/foreach}
+										</select>
+									</div>
+								</div>
+								<div class="row">
+									<div class="col">
+										<button class="btn btn-success save-scheme w-100" data-js="click"><span class="fas fa-save"></span> {\App\Language::translate('LBL_SAVE_SCHEME','Settings:PDF')}</button>
```
Comment (mariuszkrzaczkowski) [inline layouts/basic/modules/Vtiger/ExportPDF.tpl]:
```suggestion
										<button class="btn btn-success js-save-scheme w-100" data-js="click"><span class="fas fa-save"></span> {\App\Language::translate('LBL_SAVE_SCHEME','Settings:PDF')}</button>
```

File: public_html/layouts/basic/modules/Vtiger/resources/PDF.js [hunk 27]
```diff
@@ -100,48 +118,108 @@ jQuery.Class("Vtiger_PDF_Js", {
...
+	 */
+	countSelectedRecords(container) {
+		return JSON.parse(container.find('[name="selectedRecords"]').val()).length;
+	},
+
+	/**
+	 * Register save scheme button click
+	 */
+	registerSaveInventoryColumnSchemeClick() {
+		this.container.find('.save-scheme').on('click', e => {
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/Vtiger/resources/PDF.js]:
```suggestion
		this.container.find('.js-save-scheme').on('click', e => {
```

File: public_html/layouts/basic/modules/Vtiger/resources/PDF.js [hunk 28]
```diff
@@ -100,48 +118,108 @@ jQuery.Class("Vtiger_PDF_Js", {
...
+			});
 
-		return selectedRecords.length;
+			const records = JSON.parse(this.container.find('[name="selectedRecords"]').val());
+			const record = parseInt(this.container.find('[name="record"]').val());
+			if (record) {
+				records.push(record);
+			}
+			const inventoryColumns = this.container.find('[name="inventoryColumns[]"]').val();
+			const params = {};
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/Vtiger/resources/PDF.js]:
```suggestion
			AppConnector.request({
			  data: {
				module: app.getModuleName(),
				mode: 'saveInventoryColumnScheme',
				action: 'PDF',
				records,
				inventoryColumns
			},
			dataType: 'json'
			};
```

File: app/TextParser/DynamicInventoryColumnsTable.php [hunk 29]
```diff
@@ -0,0 +1,50 @@
...
+ * @license YetiForce Public License 3.0 (licenses/LicenseEN.txt or yetiforce.com)
+ * @author Rafal Pospiech <r.pospiech@yetiforce.com>
+ */
+
+namespace App\TextParser;
+
+/**
+ * Products table class.
+ *
+ * @copyright YetiForce Sp. z o.o
```
Comment (mariuszkrzaczkowski) [inline app/TextParser/DynamicInventoryColumnsTable.php]: remove

File: app/TextParser/DynamicInventoryColumnsTable.php [hunk 30]
```diff
@@ -0,0 +1,50 @@
...
+ * @license   YetiForce Public License 3.0 (licenses/LicenseEN.txt or yetiforce.com)
+ * @author    Mariusz Krzaczkowski <m.krzaczkowski@yetiforce.com>
+ */
+class DynamicInventoryColumnsTable extends Base
+{
+	/** @var string Class name */
+	public $name = 'LBL_DYNAMIC_INVENTORY_TABLE';
+
+	/** @var mixed Parser type */
```
Comment (mariuszkrzaczkowski) [inline app/TextParser/DynamicInventoryColumnsTable.php]:
```suggestion
	/** @var string Parser type */
```

**Pull request merged**